sklearn/preprocessing/data.py

```diff
@@ -138,8 +139,39 @@ def scale(X, axis=0, with_mean=True, with_std=True, copy=True):
         Xr = np.rollaxis(X, axis)
         if with_mean:
             Xr -= mean_
+            mean_1 = Xr.mean(axis=0)
+            # Verify that mean_1 is 'close to zero'. If X contains very
+            # large values, mean_1 can also be very large, due to a lack of
+            # precision of mean_. In this case, a pre-scaling of the
+            # concerned feature is efficient, for instance by its mean or
+            # maximum.
+            if isinstance(mean_1, np.ndarray):
+                if not np.allclose(mean_1, np.zeros(np.shape(mean_1)[0])):
+                    raise ValueError(
+                        "Centering failed. Dataset may contain too"
+                        "large values. You may need to prescale"
+                        "your features.")
+            else:
+                if not isclose(mean_1, 0.):
+                    raise ValueError(
+                        "Centering failed. Dataset may contain"
+                        "too large values. You may need to prescale"
+                        "your features.")
         if with_std:
             Xr /= std_
+            if with_mean:
+                mean_2 = Xr.mean(axis=0)
+                # If mean_2 is not 'close to zero', it comes from the fact that
+                # std_ is very small so that mean_2 = mean_1/std_ > 0, even if
+                # mean_1 was close to zero. The problem is thus essentially due
+                # to the lack of precision of mean_. A solution is then to
+                # substract the mean again:
+                if isinstance(mean_2, np.ndarray):
+                    if not np.allclose(mean_2, np.zeros(np.shape(mean_2)[0])):
```

Comment:
same as above, you can just use ``np.allclose(mean_2, 0)`` and I don't see why you'd nee d the if.